AlphaWallet/Common/Services/ContentLoader.swift

```diff
@@ -0,0 +1,233 @@
+                    guard let preview = preview, let pngData = preview.pngData() else { return .done(.video(video)) }
+
+                    let video = ContentLoader.Video(url: video.url, preview: preview)
+                    try? cache.set(data: pngData, for: ImageCacheKey.cacheKey(for: video.url, prefix: .videoPreview))
+
+                    return .done(.video(video))
+                }.replaceError(with: .done(.video(video)))
+                .setFailureType(to: ContentLoader.ContentLoaderError.self)
+                .eraseToAnyPublisher()
+        }
+    }
+}
+
+final class ContentLoader {
+    private let queue = DispatchQueue(label: "org.alphawallet.swift.imageLoader.processingQueue", qos: .utility)
+    private var publishers: [URLRequest: LoadContentPublisher]
+    private let session: URLSession
+    private let cache: ContentCacheStorage
+    private let decoder: ImageDecoder
+    private let interceptor: ContentLoaderResponseInterceptor
+
+    init(publishers: [URLRequest: LoadContentPublisher] = [:],
+         session: URLSession,
+         cache: ContentCacheStorage,
+         decoder: ImageDecoder = ImageDecoder(),
+         interceptor: ContentLoaderResponseInterceptor) {
+
+        self.decoder = decoder
+        self.interceptor = interceptor
+        self.cache = cache
+        self.publishers = publishers
+        self.session = session
+    }
+
+    public func fetch(_ url: URL) -> LoadContentPublisher {
+        let request = URLRequest(url: url)
+        return fetch(request)
+    }
+
+    public func fetch(_ urlRequest: URLRequest) -> LoadContentPublisher {
+        return buildFetchPublisher(urlRequest)
+            .prepend(.loading)
+            .eraseToAnyPublisher()
+    }
+
+    private func buildFetchPublisher(_ urlRequest: URLRequest) -> LoadContentPublisher {
+        guard let url = urlRequest.url else { return .fail(.invalidUrl) }
+
+        if let data = try? cache.value(for: ImageCacheKey.cacheKey(for: url, prefix: .raw)), let data = try? decoder.decode(data: data) {
+            switch data {
+            case .image, .svg:
+                return .just(.done(data))
+            case .video(let video):
+                let preview = try? cache.value(for: ImageCacheKey.cacheKey(for: url, prefix: .videoPreview)).flatMap { UIImage(data: $0) }
+                let video = Video(url: video.url, preview: preview)
+
+                return .just(.done(.video(video)))
+            }
+        }
+
+        return Just(urlRequest)
+            .receive(on: queue)
+            .setFailureType(to: ContentLoaderError.self)
+            .flatMap { [weak self, queue, session, cache, decoder, interceptor] urlRequest -> LoadContentPublisher in
+                if let publisher = self?.publishers[urlRequest] {
+                    return publisher
+                } else {
+                    let publisher = session.dataTaskPublisher(for: urlRequest)
+                        .receive(on: queue)
```

Comment:
Is the `.receive(on: queue)` a few lines above this needed? Since it's called here before work starts. Might need to double check access to `publishers` no matter whether we remove the `receive(on:)` call.